crypto/rand/drbg_lib.c

```diff
@@ -90,6 +90,21 @@ static RAND_DRBG *drbg_private;
  * |randomness| argument). This will immediately reseed the <master> DRBG.
  * The <public> and <private> DRBG will detect this on their next generate
  * call and reseed, pulling randomness from <master>.
+ *
+ * LOCKING
+ *
+ * The three shared DRBGs are intended to be used concurrently, so they
+ * support locking by default. It is the callers responsibility to wrap
+ * calls to functions like RAND_DRBG_generate() which modify the DRBGs
+ * internal state with calls to RAND_DRBG_lock() and RAND_DRBG_unlock().
+ * The functions RAND_bytes() and RAND_priv_bytes() take the locks
+ * automatically, so using the RAND api is thread safe as before.
+ *
+ * All other DRBG instances don't have locking enabled by default, because
+ * they are intendended to be used by a single thread. If it is desired,
+ * locking can be enabled using RAND_DRBG_enable_locking(). However, instead
+ * of accessing a single DRBG instance concurrently from different threads,
+ * it is recommended to instantiate a separate DRBG instance per thread.
```

Comment:
Perhaps we should explicitly note that any DRBG that is a "parent" for another DRBG must have locking enabled.

Reply:
Well, this is not necessarily the case. If parent and children are owned by the same single thread, then no locking is necessary. I think it is sufficient to make a general statement that the instances are not thread safe unless locking is enabled. For the rest, common sense w.r.t. accessing thread unsafe objects applies.

The only thing I explicitely enforce: If the user calls `RAND_drbg_enable_locking()` for a DRBG instance (which implicitely implies his intent to use it concurrently) and locking of the DRBG's parent is not enabled, then `RAND_drbg_enable_locking()` fails with an error message.